src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.cs

```diff
@@ -198,27 +171,75 @@ internal ItemManager(RecentItemsManager recentItemsManager)
                 initialRoslynTriggerKind,
                 filterReason,
                 data.Trigger.Character,
-                initialListOfItemsToBeIncluded,
-                highlightedList,
+                extendedFilterResults,
                 completionHelper,
                 hasSuggestedItemOptions);
         }
 
+        private bool TryCreateExtendedFilterResult(
+            CompletionHelper completionHelper, VSCompletionItem item,
+            string filterText, CompletionTriggerKind initialRoslynTriggerKind,
+            CompletionFilterReason filterReason, bool highlightMatchingPortions,
+            out ExtendedFilterResult extendedFilterResult)
+        {
+            if (!item.Properties.TryGetProperty(CompletionSource.RoslynItem, out RoslynCompletionItem roslynItem))
+            {
+                roslynItem = RoslynCompletionItem.Create(
+                    displayText: item.DisplayText,
+                    filterText: item.FilterText,
+                    sortText: item.SortText,
+                    displayTextSuffix: item.Suffix);
+            }
+
+            var matchesFilterText = MatchesFilterText(
+                completionHelper, roslynItem, filterText, initialRoslynTriggerKind, filterReason,
+                _recentItemsManager.RecentItems, highlightMatchingPortions, out var matchedSpans);
+
+            // Either the item matched the filter text, or, we'll still keep it in the list
+            // if one of two things is true:
+            //
+            //  1. The user has only typed a single character.  In this case they might
+            //     have just typed the character to get completion.  Filtering out items
+            //     here is not desirable.
+            //
+            //  2. They brough up completion with ctrl-j or through deletion.  In these
+            //     cases we just always keep all the items in the list.
+            if (matchesFilterText ||
+                initialRoslynTriggerKind == CompletionTriggerKind.Deletion ||
+                initialRoslynTriggerKind == CompletionTriggerKind.Invoke ||
+                filterText.Length <= 1)
+            {
+                if (matchedSpans.IsEmpty && highlightMatchingPortions)
```

Comment:
Please explain the scenario when
1. `matchedSpans` comes back empty from `MatchesFilterText`
2. `completionHelper.GetHighlightedSpans` sets it to something non-empty

If that's not going to happen, remove this if statement

Reply:
Looks like both `MatchesFilterText` and `GetHighlightedSpans` call `MatchesPattern`, so the following call to `GetHighlightedSpans` appears to be redundant. Or perhaps I've missed something!

Reply:
This was for the case if Deletion where we escape early from MatchesFilterText. I moved this fork inside MatchesFilterText

---
In reply to: [265360835](https://github.com/dotnet/roslyn/pull/33734#discussion_r265360835) [](ancestors = 265360835)